Added detection for untrusted domains in script content - edited existing library and added new query and tests ("the Polyfill PR") (#16886)
github/codeql: javascript/ql/src/Security/CWE-830/CUSTOMIZING.md

Reviewer: I suggest that we move this information into the [Customizing Library Models for JavaScript — CodeQL](https://codeql.github.com/docs/codeql-language-guides/customizing-library-models-for-javascript/) article as a practical example. Alternatively, we could create a new article in https://github.com/github/codeql/tree/main/docs/codeql/codeql-language-guides, which could be extended with future examples.

This would require converting the format to RST, but the content is fairly simple so this shouldn't be too much of an overhead and we can provide support if needed.

> # Functionaility from untrusted source/domain - customizations

Reviewer: Typo in the first word, but perhaps a task focused title is better?

```suggestion
# Extending the library list of untrusted sources and domains
```

You can extend the behavior of the `js/functionality-from-untrusted-source` and `js/functionality-from-untrusted-domain` queries using [CodeQL data extensions](https://codeql.github.com/docs/codeql-language-guides/customizing-library-models-for-javascript/).

> This allows you to require Subresource Integrity (SRI) checks on specific content delivery network (CDN) hostnames, and add additional domains to warn on, respectively.

Reviewer: I'm not sure that "respectively" adds anything in this case.

As a non-expert, it seems to me as if step one is adding more domains and step 2 is requiring SRI check, so I'd tend to suggest swapping the sentence around.

```suggestion
This allows you to add additional domains to warn users about and to require Subresource Integrity (SRI) checks on specific content delivery network (CDN) hostnames.
```

For example, this YAML model can be used inside a CodeQL model pack to alert on uses of `example.com` in imported functionality, extending the `js/functionality-from-untrusted-domain` query:

```yaml
extensions:
  - addsTo:
      pack: codeql/javascript-all
      extensible: untrustedDomain
    data:
      - ["example.com"]
```

To add new hostnames that always require SRI checking, this YAML model can be used to require SRI on `cdn.example.com`, extending the `js/functionality-from-untrusted-source` query:

```yaml
extensions:
  - addsTo:
      pack: codeql/javascript-all
      extensible: isCdnDomainWithCheckingRequired
    data:
      - ["cdn.example.com"]
```

> A suitable [model pack](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/creating-and-working-with-codeql-packs#creating-a-codeql-model-pack) might be:

Reviewer: Since we include this reference in the references section at the end, we can probably safely omit it here for simplicity:

```suggestion
You would create a model pack with this information using metadata similar to that in the example below:
```
